homeassistant/helpers/service.py

```diff
@@ -112,3 +116,57 @@ def extract_entity_ids(hass, service_call, expand_group=True):
             return [service_ent_id]
 
         return service_ent_id
+
+
+@asyncio.coroutine
+@bind_hass
+def async_get_all_descriptions(hass):
+    """Return descriptions (i.e. user documentation) for all service calls."""
+    yaml_cache = {}
+
+    if SERVICE_DESCRIPTION_CACHE not in hass.data:
+        hass.data[SERVICE_DESCRIPTION_CACHE] = {}
+
+    def get_description(hass, domain, service):
+        """Return the description for a single service call."""
+        cache_key = "{}.{}".format(domain, service)
+        description = hass.data[SERVICE_DESCRIPTION_CACHE].get(cache_key)
+
+        if description is None:
+            import homeassistant.components as components
+            catch_all_path = path.dirname(components.__file__)
+
+            if domain == ha.DOMAIN:
+                component_path = catch_all_path
+            else:
+                component_path = path.dirname(get_component(domain).__file__)
+            yaml_file = path.join(component_path, 'services.yaml')
+
+            if yaml_file not in yaml_cache:
+                try:
+                    yaml_cache[yaml_file] = load_yaml(yaml_file)
+                except FileNotFoundError:
+                    yaml_cache[yaml_file] = {}
+
+            if component_path == catch_all_path:
+                domain_services = yaml_cache[yaml_file].get(domain, {})
+            else:
+                domain_services = yaml_cache[yaml_file]
+            description = domain_services.get(service, {})
+
+            hass.data[SERVICE_DESCRIPTION_CACHE][cache_key] = description
+
+        return {
+            'description': description.get('description', ''),
+            'fields': description.get('fields', {})
+        }
+
+    services = hass.services.async_services()
+    descriptions = {}
+    for domain in services:
+        descriptions[domain] = {}
+        for service in services[domain]:
+            descriptions[domain][service] = yield from hass.async_add_job(
```

Comment:
Your code would be a lot faster if you test if the key exists in the cache inside the async context (before this line). No need to yield and wait for a thread to run our job.